File: backend/routers/vision.py
import logging
import threading
import time
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from sqlmodel import Session, select

from .. import config
from ..database import get_session
from ..dependencies.auth import get_current_user
from ..dependencies.patient import default_patient, get_optional_patient
from ..models import AIAnalysis, Instance, Patient, Series, User
from ..services.dicom_service import store_patient_dicom
from ..services.explanation import cxr_explanation
from ..services.vision import analyses, results
from ..services.vision.errors import (InvalidImageError, UnknownTargetError, UnsupportedImageError,
                                      VisionModelError, VisionProviderUnavailable)
from ..services.vision.provider import get_vision_provider, provider_key
from ..services.vision.schemas import VisionScreenResponse
from .medical_data import log_action

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/vision", tags=["Vision AI"])

_CHUNK = 1024 * 1024
EXPLANATION_UNAVAILABLE = ("AI explanation is temporarily unavailable. The screening result and "
                           "visual explanation remain available for clinical review.")


async def _read_limited(file: UploadFile, limit: int) -> bytes:
    """Read the upload into memory, rejecting it as soon as it exceeds ``limit``."""
    chunks, total = [], 0
    while True:
        chunk = await file.read(_CHUNK)
        if not chunk:
            break
        total += len(chunk)
        if total > limit:
            raise HTTPException(status_code=413,
                                detail=f"File too large: limit is {limit // (1024 * 1024)} MiB")
        chunks.append(chunk)
    return b"".join(chunks)


async def run_screen(data: bytes, target: Optional[str], user: User):
    """Run the configured vision provider; returns (response with result_id, provider name)."""
    started = time.perf_counter()
    try:
        provider = get_vision_provider()
        result = await run_in_threadpool(provider.screen, data, target or None)
    except VisionProviderUnavailable as exc:
        logger.warning("Vision provider unavailable: %s", exc)
        raise HTTPException(status_code=503, detail="Vision service is not available")
    except ImportError:
        # The local provider imports torch lazily; the backend runs without it.
        logger.exception("Vision stack is not installed")
        raise HTTPException(status_code=503, detail="Vision service is not available")
    except UnsupportedImageError as exc:
        raise HTTPException(status_code=415 if exc.media_type else 422, detail=str(exc))
    except InvalidImageError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    except UnknownTargetError as exc:
        raise HTTPException(status_code=422, detail=str(exc))
    except VisionModelError:
        logger.exception("Vision model unavailable")
        raise HTTPException(status_code=503, detail="Vision model is not available")
    except Exception:
        logger.exception("Vision screening failed")
        raise HTTPException(status_code=500, detail="Vision screening failed")

    scores = {f.pathology: f.score for f in result.findings}
    result.result_id = results.register(user.id, scores, result.primary_finding.pathology,
                                        result.model.weights, result.model.weight_sha256)
    logger.info("vision_screen provider=%s format=%s target=%s ms=%.0f",
                provider.name, result.input.format, result.explanation.target_pathology,
                (time.perf_counter() - started) * 1000)
    return result, provider.name


def log_screen(session: Session, user: User, result: VisionScreenResponse, patient: Optional[Patient] = None) -> None:
    log_action(session, user.id, "vision_screen", {
        "input_format": result.input.format,
        "weights": result.model.weights,
        "target_pathology": result.explanation.target_pathology,
    }, patient_id=patient.id if patient else None)
    session.commit()


def attach_analysis(result: VisionScreenResponse, analysis: AIAnalysis, patient: Patient, session: Session) -> None:
    summary = analyses.summary(session, analysis)
    result.analysis_id = analysis.uid
    result.patient_id = patient.uid
    result.study_instance_uid = summary["study_instance_uid"]
    result.sop_instance_uid = summary["sop_instance_uid"]
    if analysis.instance_id:
        instance = session.get(Instance, analysis.instance_id)
        series = session.get(Series, instance.series_id) if instance else None
        result.series_instance_uid = series.series_instance_uid if series else None


@router.post("/screen", response_model=VisionScreenResponse)
async def screen(
    file: UploadFile = File(..., description="Chest radiograph: PNG, JPEG, or DICOM"),
    target: Optional[str] = Form(None, description="Optional Grad-CAM target; must be a model output"),
    user: User = Depends(get_current_user),
    session: Session = Depends(get_session),
    save: Annotated[bool, Form(description="Save the result (and a DICOM input) to the active patient")] = False,
    analysis_id: Annotated[Optional[str], Form(description="Saved analysis this run re-targets")] = None,
    active: Annotated[Optional[Patient], Depends(get_optional_patient)] = None,
):
    """Chest radiograph screening assistance with a Grad-CAM visual explanation.

    By default the upload is processed in memory and not stored. With ``save`` the result is
    kept as an analysis of the active patient and a DICOM input is stored unchanged as that
    patient's imaging study. Scores are model scores, not diagnoses; every response carries
    the safety notice.
    """
    data = await _read_limited(file, config.VISION_MAX_UPLOAD_BYTES)
    if not data:
        raise HTTPException(status_code=400, detail="Empty upload")
    result, provider_name = await run_screen(data, target or None, user)

    if not (save or analysis_id):
        log_screen(session, user, result)
        return result
    patient = active or default_patient(session, user)
    if analysis_id:
        analysis = session.exec(select(AIAnalysis).where(AIAnalysis.uid == analysis_id, AIAnalysis.owner_id == user.id,
                                                         AIAnalysis.patient_id == patient.id)).first()
        if analysis is None:
            raise HTTPException(status_code=404, detail="Analysis not found")
        analysis.selected_target = result.explanation.target_pathology
        session.add(analysis)
        session.commit()
    else:
        study = instance = None
        if result.input.format == "dicom":
            stored = store_patient_dicom(session, user, patient, data, file.filename or "screening.dcm")
            study, instance = stored.study, stored.instance
        analysis = analyses.record(session, user, patient, result, data, provider_name, study, instance)
    attach_analysis(result, analysis, patient, session)
    log_screen(session, user, result, patient)
    return result


class ExplanationRequest(BaseModel):
    result_id: str = Field(..., min_length=8, max_length=64, pattern=r"^[A-Za-z0-9_-]+$")
    target: str = Field(..., min_length=1, max_length=64)


@router.post("/explanations", response_model=cxr_explanation.TextExplanationResponse)
async def explain(
    body: ExplanationRequest,
    user: User = Depends(get_current_user),
    session: Session = Depends(get_session),
):
    """Language-model explanation of one model output of a screening result.

    Uses only the structured result stored by /screen (never the image). If the
    text AI is unavailable the screening result is unaffected; this returns 503.
    """
    try:
        response = await run_in_threadpool(cxr_explanation.explain, body.result_id, user.id, body.target)
    except cxr_explanation.ExplanationNotFound:
        raise HTTPException(status_code=404, detail="Screening result not found or expired")
    except cxr_explanation.ExplanationUnavailable:
        raise HTTPException(status_code=503, detail=EXPLANATION_UNAVAILABLE)
    except Exception:
        logger.exception("Explanation failed")
        raise HTTPException(status_code=503, detail=EXPLANATION_UNAVAILABLE)
    if not response.cached:
        log_action(session, user.id, "vision_explanation", {"target_pathology": response.target_pathology})
        session.commit()
        analyses.save_text_explanation(session, user.id, body.result_id, response.target_pathology,
                                       response.model_dump(mode="json"))
    return response


_status_cache = {"at": 0.0, "value": None}
_status_lock = threading.Lock()
STATUS_CACHE_SECONDS = 10.0


@router.get("/status")
async def status():
    """Readiness of the vision service (no credentials or infrastructure details)."""
    with _status_lock:
        cached = _status_cache["value"]
        if cached is not None and time.monotonic() - _status_cache["at"] < STATUS_CACHE_SECONDS:
            return cached
    key = provider_key()
    try:
        provider = get_vision_provider()
        info = await run_in_threadpool(provider.status)
    except VisionProviderUnavailable:
        info = {"configured": False, "model_ready": False}
    value = {
        "provider": key if key in ("local", "cloud") else "unknown",
        "provider_configured": bool(info.get("configured")),
        "model_ready": bool(info.get("model_ready")),
        "accelerator": info.get("accelerator") if info.get("accelerator") in ("gpu", "cpu") else None,
        "status": ("ready" if info.get("model_ready")
                   else "unavailable" if not info.get("configured") or info.get("reachable") is False
                   else "loading" if key == "cloud" or config.VISION_PRELOAD
                   else "standby"),  # local model loads on first request
    }
    with _status_lock:
        _status_cache.update(at=time.monotonic(), value=value)
    return value

File: frontend/src/lib/vision.ts
/**
 * Chest X-ray AI screening client — POST /api/v1/vision/screen.
 *
 * Types mirror backend/services/vision/schemas.py (VisionScreenResponse).
 * The browser only talks to the HoloMed backend (session cookie via api.ts);
 * image bytes are sent once per request and never stored client-side.
 */

import { api, ApiError } from '@/lib/api';

// ─── Response schema (backend/services/vision/schemas.py) ─────────────────────

export interface VisionModelInfo {
  name: string;
  architecture: string;
  weights: string;
  weight_sha256: string;
  targets: number;
  target_list: string[];
  device: string;
  input_size: number;
  score_semantics: string;
}

export interface VisionInputInfo {
  format: 'png' | 'jpeg' | 'dicom';
  width: number;
  height: number;
  source_mode: string;
  bits_stored: number | null;
  modality: string | null;
  transfer_syntax: string | null;
  preprocessing: string[];
}

export interface VisionFinding {
  pathology: string;
  score: number;
}

export interface EncodedImage {
  media_type: 'image/png';
  encoding: 'base64';
  width: number;
  height: number;
  data: string;
}

export interface VisionExplanation {
  method: 'Grad-CAM';
  target_pathology: string;
  target_score: number;
  target_layer: string;
  description: string;
  original: EncodedImage;
  heatmap: EncodedImage;
  overlay: EncodedImage;
}

export interface VisionTiming {
  preprocessing_ms: number;
  inference_ms: number;
  gradcam_ms: number;
  rendering_ms: number;
  total_ms: number;
}

export interface VisionSafety {
  message: string;
  requires_clinical_review: boolean;
}

export interface VisionScreenResponse {
  model: VisionModelInfo;
  input: VisionInputInfo;
  primary_finding: VisionFinding;
  findings: VisionFinding[];
  explanation: VisionExplanation;
  timing: VisionTiming;
  inferred_at: string;
  safety: VisionSafety;
  /** Short-lived handle for requesting a text explanation of this result. */
  result_id: string | null;
  /** Set when the result was saved to a patient (and, for DICOM, the stored study it belongs to). */
  analysis_id?: string | null;
  patient_id?: string | null;
  study_instance_uid?: string | null;
  series_instance_uid?: string | null;
  sop_instance_uid?: string | null;
}

// ─── Text explanation (POST /api/v1/vision/explanations) ──────────────────────
// Mirrors backend/services/explanation/cxr_explanation.py (TextExplanationResponse).

export interface ExplanationSections {
  summary: string;
  finding_explanation: string;
  score_explanation: string;
  gradcam_explanation: string;
  limitations: string[];
  clinical_review: string;
}

export interface TextExplanationResponse {
  result_id: string;
  target_pathology: string;
  model_score: number;
  is_primary_finding: boolean;
  source: 'language-model';
  text_model: string;
  generated_at: string;
  cached: boolean;
  explanation: ExplanationSections;
  safety: VisionSafety;
}

/** A screening response plus the browser-measured round-trip time. */
export interface ScreeningRun {
  response: VisionScreenResponse;
  roundTripMs: number;
}

// ─── Client-side validation (the backend re-validates everything) ─────────────

export const MAX_UPLOAD_BYTES = 50 * 1024 * 1024; // matches VISION_MAX_UPLOAD_BYTES default
export const ACCEPT_ATTR = '.png,.jpg,.jpeg,.dcm,.dicom,image/png,image/jpeg,application/dicom';
const REQUEST_TIMEOUT_MS = 90_000;

export type DetectedFormat = 'png' | 'jpeg' | 'dicom';

/** Identify PNG / JPEG / DICOM Part 10 from the file signature. */
export async function detectFormat(file: File): Promise<DetectedFormat | null> {
  const head = new Uint8Array(await file.slice(0, 132).arrayBuffer());
  const png = [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a];
  if (png.every((b, i) => head[i] === b)) return 'png';
  if (head[0] === 0xff && head[1] === 0xd8 && head[2] === 0xff) return 'jpeg';
  if (head.length >= 132 && String.fromCharCode(...head.slice(128, 132)) === 'DICM') return 'dicom';
  return null;
}

export async function validateUpload(file: File): Promise<{ format: DetectedFormat } | { error: string }> {
  if (file.size === 0) return { error: 'The selected file is empty.' };
  if (file.size > MAX_UPLOAD_BYTES) return { error: 'The file is larger than the 50 MB upload limit.' };
  const format = await detectFormat(file);
  if (!format) {
    return { error: 'Unsupported file type. Select a PNG, JPEG, or DICOM (.dcm) chest X-ray.' };
  }
  return { format };
}

// ─── Request ──────────────────────────────────────────────────────────────────

export interface SaveOptions {
  /** Save the result (and a DICOM input) to the active patient. */
  save?: boolean;
  /** Re-target an already saved analysis instead of saving a new one. */
  analysisId?: string | null;
}

async function timed(call: (signal: AbortSignal) => Promise<VisionScreenResponse>): Promise<ScreeningRun> {
  const controller = new AbortController();
  const timer = window.setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
  const started = performance.now();
  try {
    const response = await call(controller.signal);
    return { response, roundTripMs: performance.now() - started };
  } finally {
    window.clearTimeout(timer);
  }
}

export async function screenChestXray(file: File, target?: string, options: SaveOptions = {}): Promise<ScreeningRun> {
  const form = new FormData();
  form.append('file', file);
  if (target) form.append('target', target);
  if (options.save) form.append('save', 'true');
  if (options.analysisId) form.append('analysis_id', options.analysisId);
  return timed((signal) => api.postMultipart<VisionScreenResponse>('/api/v1/vision/screen', form, { signal }));
}

/** Screen a study already stored for the patient (no re-upload); saved unless re-targeting `analysisId`. */
export async function screenStoredStudy(patientId: string, studyUid: string, target?: string,
  analysisId?: string | null): Promise<ScreeningRun> {
  const params = new URLSearchParams();
  if (target) params.set('target', target);
  if (analysisId) params.set('analysis_id', analysisId);
  const query = params.toString();
  const path = `/api/v1/patients/${patientId}/imaging/${encodeURIComponent(studyUid)}/screen${query ? `?${query}` : ''}`;
  return timed((signal) => api.post<VisionScreenResponse>(path, {}, { signal }));
}

export interface SavedAnalysis {
  id: string;
  study_instance_uid: string | null;
  primary_pathology: string;
  primary_score: number;
  selected_target: string | null;
  created_at: string;
  response: VisionScreenResponse;
  text_explanations: Record<string, TextExplanationResponse>;
}

export function getSavedAnalysis(patientId: string, analysisId: string): Promise<SavedAnalysis> {
  return api.get<SavedAnalysis>(`/api/v1/patients/${patientId}/analyses/${analysisId}`);
}

export function listSavedAnalyses(patientId: string): Promise<Omit<SavedAnalysis, 'response' | 'text_explanations'>[]> {
  return api.get(`/api/v1/patients/${patientId}/analyses`);
}

// ─── Error mapping (never surface raw server text) ────────────────────────────

export interface ScreeningError {
  title: string;
  message: string;
  kind: 'auth' | 'input' | 'unavailable' | 'network' | 'timeout' | 'unknown';
}

export function describeScreeningError(err: unknown): ScreeningError {
  if (err instanceof DOMException && err.name === 'AbortError') {
    return { kind: 'timeout', title: 'Request timed out', message: 'The screening request took too long. Please try again.' };
  }
  if (err instanceof ApiError) {
    const detail = typeof err.detail === 'string' ? err.detail.toLowerCase() : '';
    switch (err.status) {
      case 401:
        return { kind: 'auth', title: 'Session expired', message: 'Your session has ended. Please sign in again to run AI screening.' };
      case 400:
        return { kind: 'input', title: 'Image could not be read', message: 'The file appears to be corrupt, incomplete, or empty. Please select another image.' };
      case 413:
        return { kind: 'input', title: 'File too large', message: 'The file is larger than the 50 MB upload limit.' };
      case 415:
        return { kind: 'input', title: 'Unsupported file type', message: 'Select a PNG, JPEG, or DICOM (.dcm) chest X-ray.' };
      case 422:
        if (detail.includes('target')) {
          return { kind: 'input', title: 'Finding not available', message: 'The selected finding is not an output of this model.' };
        }
        return {
          kind: 'input',
          title: 'Image not supported for screening',
          message: 'Screening requires a single-frame grayscale chest radiograph (DICOM CR/DX, PNG, or JPEG) at least 64 px on each side.',
        };
      case 500:
      case 502:
      case 503:
      case 504:
        return { kind: 'unavailable', title: 'Screening unavailable', message: 'Screening service is temporarily unavailable. Please try again.' };
      default:
        return { kind: 'unknown', title: 'Screening failed', message: 'The screening request could not be completed. Please try again.' };
    }
  }
  if (err instanceof TypeError) {
    return { kind: 'network', title: 'Backend unreachable', message: 'Cannot reach the HoloMed backend. Check your connection and that the server is running, then try again.' };
  }
  return { kind: 'unknown', title: 'Screening failed', message: 'The screening request could not be completed. Please try again.' };
}

// ─── Vision service status (GET /api/v1/vision/status) ────────────────────────

export interface VisionStatus {
  provider: 'local' | 'cloud' | 'unknown';
  provider_configured: boolean;
  model_ready: boolean;
  accelerator: 'gpu' | 'cpu' | null;
  status: 'ready' | 'loading' | 'standby' | 'unavailable';
}

export function getVisionStatus(): Promise<VisionStatus> {
  return api.get<VisionStatus>('/api/v1/vision/status');
}

/** Human label for where inference runs; never includes URLs or credentials. */
export function providerLabel(status: VisionStatus | null): string {
  if (!status) return 'Checking…';
  if (status.provider === 'cloud') return 'Managed GPU';
  if (status.provider === 'local') return status.accelerator === 'cpu' ? 'Local CPU' : 'Local GPU';
  return 'Unavailable';
}

/** "cuda:0 (NVIDIA GeForce RTX 3070 Ti)" → "NVIDIA GeForce RTX 3070 Ti"; "cpu" → "CPU". */
export function deviceLabel(device: string): string {
  const match = device.match(/\(([^)]+)\)/);
  if (match) return match[1];
  return device.toLowerCase() === 'cpu' ? 'CPU' : device;
}

const EXPLANATION_TIMEOUT_MS = 120_000;

/** Request a language-model explanation for one model output of a screening result. */
export async function explainFinding(resultId: string, target: string): Promise<TextExplanationResponse> {
  const controller = new AbortController();
  const timer = window.setTimeout(() => controller.abort(), EXPLANATION_TIMEOUT_MS);
  try {
    return await api.post<TextExplanationResponse>(
      '/api/v1/vision/explanations',
      { result_id: resultId, target },
      { signal: controller.signal },
    );
  } finally {
    window.clearTimeout(timer);
  }
}

export const EXPLANATION_UNAVAILABLE_TEXT =
  'AI explanation is temporarily unavailable. The screening result and visual explanation remain available for clinical review.';

export function describeExplanationError(err: unknown): { message: string; retryable: boolean; auth: boolean } {
  if (err instanceof ApiError && err.status === 401) {
    return { message: 'Your session has ended. Please sign in again.', retryable: false, auth: true };
  }
  if (err instanceof ApiError && err.status === 404) {
    return {
      message: 'This screening result has expired. Run AI screening again to request an explanation.',
      retryable: false,
      auth: false,
    };
  }
  return { message: EXPLANATION_UNAVAILABLE_TEXT, retryable: true, auth: false };
}

export function imageDataUrl(img: EncodedImage): string {
  return `data:${img.media_type};base64,${img.data}`;
}
